sbi/inference/abc/abc_base.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import annotations
+from abc import ABC
+
+import torch
+from typing import Callable, Union
+
+from sbi.simulators.simutils import simulate_in_batches
+from sbi.user_input.user_input_checks import process_x, prepare_for_sbi
+from numpy import ndarray
+from torch import Tensor
+
+
+class ABCBASE(ABC):
+    def __init__(
+        self,
+        simulator: Callable,
+        prior,
+        x_o: Union[Tensor, ndarray],
+        distance: Union[str, Callable] = "l2",
+        num_workers: int = 1,
+        simulation_batch_size: int = 1,
+        show_progress_bars: bool = True,
+    ) -> None:
+        r"""Base class for Approximate Bayesian Computation methods.
+
+        Args:
+            simulator: A function that takes parameters $\theta$ and maps them to
+                simulations, or observations, `x`, $\mathrm{sim}(\theta)\to x$. Any
+                regular Python callable (i.e. function or class with `__call__` method)
+                can be used.
+            prior: A probability distribution that expresses prior knowledge about the
+                parameters, e.g. which ranges are meaningful for them. Any
+                object with `.log_prob()`and `.sample()` (for example, a PyTorch
+                distribution) can be used.
+            x_o: Observed data.
+            distance: Distance function to compare observed and simulated data. Can be  
+                a custom function or one of "l1", "l2", "mse".
+            num_workers: Number of parallel workers to use for simulations.
+            simulation_batch_size: Number of parameter sets that the simulator
+                maps to data x at once. If None, we simulate all parameter sets at the
+                same time. If >= 1, the simulator has to process data of shape
+                (simulation_batch_size, parameter_dimension).
+            show_progress_bars: Whether to show a progressbar during simulation and
+                sampling.
+        """
+
+        # User input checks.
+        simulator, prior, x_shape = prepare_for_sbi(simulator, prior)
```

Comment:
this should not be here I think. For the flexible interface, we require the user to run this themselves.